src/librbd/io/ImageRequestWQ.cc

```diff
@@ -452,6 +452,20 @@ void ImageRequestWQ::process(ImageRequest<> *req) {
   ldout(cct, 20) << "ictx=" << &m_image_ctx << ", "
                  << "req=" << req << dendl;
 
+  if (m_image_ctx.get_max_iops()) { //if we set limit value
+    utime_t curr_L = m_image_ctx.qos_status->get_prev_limit();
+    utime_t now = ceph_clock_now();
+
+    curr_L += (double) (1.0 / m_image_ctx.get_max_iops());
+    curr_L = (curr_L > now ? curr_L : now);
+    if (curr_L > now) {
+      ldout(cct, 20) << "we are ahead by " << curr_L - now << ", sleep" << dendl;
+      utime_t delay = curr_L - now;
+      delay.sleep();
```

Comment:
Don't sleep --- this will block other operations, not just IO. Instead, you will need to (1) move this logic to ```_void_dequeue```, (2) set a timer callback to wake up when you need to delay, (3) flag that IO is blocked due to throttling, (4) return nullptr to indicate to the thread pool there is no more work, and (5) once that timer expires, you can reset the flag and signal the thread pool to wake up.